add I2S camera driver

rursprung opened the pull request:
this is an attempt at rebasing #1646 to the latest version. all credits go to @Dominaezzz for the original implementation.
this isn't ready for actual review yet as it hasn't been tested yet (the first attempt at testing it didn't work on ESP32-CAM) and there are some open TODOs (see the commit-msg).

note: there'll be force-pushes to this branch until this PR is ready.

---

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [ ] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [ ] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [ ] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [ ] My changes are in accordance to the [esp-rs developer guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/DEVELOPER-GUIDELINES.md)

#### Extra:
- [ ] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### Pull Request Details 📖
see the commit-msg for further details.

#### Description
see the commit-msg for further details.

#### Testing
Describe how you tested your changes.
**TODO**

bugadani commented on esp-hal/src/i2s/master/mod.rs:
```diff
@@ -107,6 +107,9 @@ use crate::{
     DriverMode,
 };
 
+//#[cfg(esp32)]
```
I'm not a fan of the organization (camera shouldn't be under the i2s master IMHO) but either way, camera needs to be gated for esp32

Dominaezzz replied:
Yeah, it's actually the slave mode. `i2s/slave/camera.rs`

bugadani replied:
Eh, just place it as `i2s/camera.rs`, we didn't put the display driver anywhere either.

rursprung replied:
IIRC i had to put it under `master` as i had to reference a bunch of things from `master` which aren't `pub(crate)` (or `pub`). but i can check on that again once i know that the code works (so far testing with ESP32-CAM wasn't successful).

(and yes, the commented-out `cfg` was an oversight which i'll fix)

Dominaezzz commented:
I ran the example and the I2C bit works. The DMA streaming bit doesn't because the CPU was too slow, suspicious.

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,379 @@
+
+impl<'d, I2S: RegisterAccess, BUF: DmaRxBuffer> CameraTransfer<'d, I2S, BUF> {
+    /// Returns true when [Self::wait] will not block.
+    pub fn is_done(&self) -> bool {
```
The booleans of this function are inverted, which is why the DMA wasn't working for me.

After fixing this function I do get a valid JPEG

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,379 @@
+    /// Returns true when [Self::wait] will not block.
+    pub fn is_done(&self) -> bool {
+        // Wait for IN_SUC_EOF (i.e. I2S_RXEOF_NUM_REG / byte count threshold)
+        if self.camera._i2s.regs().int_raw().read().in_suc_eof().bit_is_set() {
+            return false;
+        }
```
Delete this check too. IN_SUC_EOF doesn't reflect done-ness in this driver.

rursprung commented:
thanks a lot for testing this! i've resolved your immediate feedbacks which should get the example running for you OOTB.
as a next step i'll try to get it working on my ESP32-CAM and then clean up the PR and mark it ready for review

rursprung commented:
i now got this to work on my ESP32-CAM board as well! the following changes are needed for that in the example:
```diff
diff --git a/qa-test/src/bin/i2s_camera.rs b/qa-test/src/bin/i2s_camera.rs
index 08d761af..cf867786 100644
--- a/qa-test/src/bin/i2s_camera.rs
+++ b/qa-test/src/bin/i2s_camera.rs
@@ -2,13 +2,13 @@
 #![feature(iter_from_coroutine)]

 //! This shows how to continuously receive data from a Camera via I2S,
-//! using a "Makerfabs ESP32 3.2" TFT LCD with Camera"
+//! using an AI Thinker ESP32-CAM board.
 //!
 //! This example reads a JPEG from an OV2640 and writes it to the console as
 //! hex.
 //!
 //! Pins used:
-//! XCLK    GPIO32
+//! XCLK    GPIO0
 //! SIOD    GPIO26
 //! SIOC    GPIO27
 //! PCLK    GPIO22
@@ -48,6 +48,11 @@ use esp_hal::{
     main,
     time::Rate,
     Blocking,
+    gpio::{
+        Output,
+        Level,
+        OutputConfig,
+    },
 };
 use esp_println::println;

@@ -62,7 +67,10 @@ fn main() -> ! {

     let cam_siod = peripherals.GPIO26;
     let cam_sioc = peripherals.GPIO27;
-    let cam_xclk = peripherals.GPIO32;
+    let cam_xclk = peripherals.GPIO0;
+
+    /// power down pin - must be pulled low for the camera to run!
+    let _cam_pwdn = Output::new(peripherals.GPIO32, Level::Low, OutputConfig::default());

     let camera = Camera::new(
         peripherals.I2S0,
```

this is still marked as a draft because i have a few TODOs left for which i do however need input:
* add missing docs for `DataFormat`
* add missing example to docs
* should the qa-test example be adapted to the ESP32-CAM instead? which board is more prevalent?

rursprung reviewed (commented):
once i have your input on this i'll update the PR and then it should be hopefully ready to merge. it'd be great if you could already review the rest of it - thanks!

rursprung commented on esp-hal/src/i2s/master/mod.rs:
```diff
@@ -834,6 +837,29 @@ mod private {
         fn din_signal(&self) -> InputSignal;
     }
 
+    pub trait ExtendedSignals: Signals {
```
CI currently complains since this generates warnings for any non-`esp32`-cfg builds

should this trait have a `#[cfg(esp32)]`? or is there a (reasonable) way to implement it for all chip types?

while the esp32s2 & esp32s3 also have e.g. the `I2S0I_WS` signal the same isn't true for e.g. `I2S0I_DATA_0`, thus i'm not sure if any other chip would support this (and just has different names for these signals)?

Dominaezzz replied:
Just cfg it for now

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
````diff
@@ -0,0 +1,369 @@
+//!
+//! ## Examples
+//! ```rust, no_run
+//! // FIXME: add example
````
Follow that pattern, this driver can be used for other purposes besides a camera.

There are plans for fully functioning examples but they won't go in the docs, they'll be stand-alone projects.

rursprung replied:
what else can it be used for (and is it correct that it's called `camera` in that case)?

Dominaezzz replied:
You could make a logic analyser with it, you can use it as a general high speed bus for receiving data.

Yeah the name is fine, camera is the primary use case. Though it's a bit inconsistent with the I2SParallel driver but that can be dealt with later once the maintainers decide on naming conventions for drivers.

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,369 @@
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum DataFormat {
+    /// FIXME: docs
```
https://github.com/espressif/esp32-camera/blob/4467667b71f22a4c7db226f24105a9350abe7a05/target/esp32/ll_cam.c#L66

You can also have a look at the TRM.

The descriptions are largely unsatisfying

rursprung replied:
`DualChannel32   = 2,` doesn't exist in the C code. is it correct that it exists here?

Dominaezzz replied:
It's "correct", it's just that esp32-camera doesn't use it.

rursprung replied:
what should the docs for `DualChannel32` say (presuming i take over the ones from esp32-camera for the others)? or should i not take the docs over anyway and add a more-or-less pointless comment which says the same as the name already says just to make the linting happy (but not actually helping any user to decide what they should pick)?

Dominaezzz replied:
I don't have a good answer, personally I'd want to test the peripheral to see what it does with the different enums, then document the observed behaviour. 

If it makes it easier you can exclude the enum entry for now

rursprung replied:
i've taken the liberty to exclude it for now. i've tried to test it, but i just ran into timeout issues whenever i tried to print everything in the example (and admittedly didn't want to invest a lot of time into this right now - it's not supported by esp32-camera and i have no need for it; i've left it with a comment so that whoever comes looking for it knows that they can just test it and add it if it works)

Dominaezzz commented on qa-test/src/bin/i2s_camera.rs:
```diff
@@ -0,0 +1,508 @@
+#![feature(iter_from_coroutine)]
+
+//! This shows how to continuously receive data from a Camera via I2S,
+//! using a "Makerfabs ESP32 3.2" TFT LCD with Camera"
```
Since this is a QA test it should be something the maintainers can easily run.
I'm not sure who has what devkit.
I know @MabezDev has a ESP32+OV2640 board but I don't which one.
I suspect @bjoernQ might have one too.

Or perhaps it's more practical for them to change the qa-test to something convenient after the PR lands

rursprung replied:
i've left it at your example for now. if one of the maintainers wants i can switch it to my ESP32-CAM example, otherwise i leave it up to them to adapt it afterwards to whichever board they have (which i can't test with anyway)

rursprung commented:
FYI: i've left this under `i2s::master` since a lot of the code is based on traits which are present in there. i'm not sure how you'd want to best refactor this with an `i2s::slave` (duplicate code? move things up to `i2s`? does that even make sense?).
note that i'm far out of my waters with this - it's the first time i'm doing anything with i2s (i just wanted to get some pictures from that camera and wanted to use rust for it since i'm always advocating for it 😅)

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,404 @@
+    }
+
+    fn stop_peripherals(&mut self) {
+        self.camera._i2s.rx_stop();
```
the DMA needs to be stopped here too

rursprung replied:
does the order matter? should i stop DMA first or I2S RX?

Dominaezzz replied:
Do what the other rx drivers do

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,404 @@
+        mut buf: BUF,
+        len: usize,
+    ) -> Result<CameraTransfer<'d, I2S, BUF>, Error> {
+        self.start_rx_transfer(&mut buf, len)?;
```
Mind just pasting the body of this function here? No point having a separate function now

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,404 @@
+        self
+    }
+
+    /// Configures the vertical sync enable pin for the camera interface.
```
typo

rursprung replied:
i don't see it? 🙈

Dominaezzz replied:
`with_h_enable` vs "vertical sync enable pin"

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,404 @@
+{
+    /// Configures the data pins for the camera interface.
+    #[allow(clippy::too_many_arguments)]
+    pub fn with_data_pins<
```
I know it's a bit tedious but these should be split out into a function per pin, per [guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/DEVELOPER-GUIDELINES.md).

rursprung replied:
doesn't that then run the (big) risk of a user accidentally missing a pin? we can't catch such an oversight and there's (probably?) never any valid reason to use this functionality with anything but the full set of data pins?
also, the current API is aligned with `lcd_cam::cam`, making it easier to move between the two. shouldn't that be valued higher (and both be refactored together at a later step by maintainers when they decide on what the APIs should look like for stabilisation)?

Dominaezzz replied:
> the current API is aligned with `lcd_cam::cam`

https://github.com/esp-rs/esp-hal/pull/3237 :smile:

rursprung replied:
as discussed [on matrix](https://matrix.to/#/!LdaNPfUfvefOLewEIM:matrix.org/$cUiQVpZ8Qi45aeLV5ch0WzeUAx4X-InBZps_TUGx6t0?via=matrix.org&via=tchncs.de&via=mozilla.org) i've now added individual `with_d[0..7]` functions in addition to `with_data_pins`

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,404 @@
+        // Reset RX unit and RX FIFO
+        self._i2s.reset_rx();
+
+        // Enable corresponding interrupts if needed
```
delete this comment

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,404 @@
+                .and_then(|_| self.rx_channel.start_transfer())?;
+        }
+
+        // set I2S_RX_STOP_EN if needed
```
It seems this bit doesn't exist in the TRM

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,404 @@
+    /// Creates a new `Camera` instance with DMA support.
+    pub fn new<CH: RxChannelFor<I2S>>(
+        i2s: impl Peripheral<P = I2S> + 'd,
+        data_format: DataFormat,
```
This probably needs to go into a `Config` struct like all the other drivers have.

rursprung replied:
what needs to be in `Config`? just `data_format`?

Dominaezzz replied:
yes

rursprung commented:
> You should be able to reference the traits in `i2s::master` from a different file. If visibility is an issue feel free to use `pub(crate)` on whatever.

this is not that simple since these traits are in `i2s::master::private` - i doubt that i should make that `pub(crate)`, that'd defeat its purpose? so either these are "master-only" (which doesn't seem to be the case) or maybe they shouldn't be in `private`?

Dominaezzz commented:
`private` means private to the crate (`esp_hal`), rather than private to the file.

Though I agree, some refactoring needs doing.

rursprung commented:
[as discussed on matrix](https://matrix.to/#/!LdaNPfUfvefOLewEIM:matrix.org/$IkI59Yphh_NdnezrTExA7Iyx9hFnTXlZ37Ni0zQ89QU?via=matrix.org&via=tchncs.de&via=mozilla.org) i've left it under `i2s::master` for the time being.

i've incorporated all your review feedback (thanks!) except for the `with_data_pins` comment - see my comment above as well as on your `lcd_cam::cam` PR (https://github.com/esp-rs/esp-hal/pull/3237#discussion_r1986709671): i think using `with_d[0..7]` methods would be a much worse API (& developer experience) compared to this or a struct (as used in `lcd_cam::cam` so far). i hope a maintainer can chip in on this discussion and then i'll adapt to whatever the outcome of it is (even if it's the ugly API 🤷)

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,406 @@
+        while !self.is_done() {}
+
+        // Stop the DMA as it doesn't know that the camera has stopped.
+        self.camera.rx_channel.stop_transfer();
```
The I2S has to be stopped too

rursprung commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,461 @@
+    }
+
+    /// Configures the d0 pin.
+    /// Use either [`with_data_pins`] *or* [`with_d0`]..[`with_d7`].
```
not sure why, but these rustdoc links don't work and i'm not sure how to fix it 😕

bugadani replied:
```suggestion
    /// Use either [`Self::with_data_pins`] *or* [`Self::with_d0`]..[`Self::with_d7`].
```

rursprung commented:
i've now created an example application which can use this: https://github.com/rursprung/esp32-cam-test/

it'd be great if this PR would get approved & merged, thanks!

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,470 @@
+            _i2s: i2s,
+            rx_channel,
+        }
+    }
```
You're missing an `apply_config` function

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,470 @@
+
+    /// Configures the d0 pin.
+    /// Use either [`Self::with_data_pins`] *or*
+    /// [`Self::with_d0`]..[`Self::with_d7`].
```
Mind exposing the remaining 8 pins? i.e. This should go from d0 to d15

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,470 @@
+        mut self,
+        mut buf: BUF,
+        len: usize,
+    ) -> Result<CameraTransfer<'d, I2S, BUF>, Error> {
```
When this fails, along with the error, the camera and the buffer should be returned.

Like `Result<CameraTransfer<'d, I2S, BUF>, (Error, Self, BUF)>`

Dominaezzz commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,470 @@
+}
+
+/// Represents the camera interface.
+pub struct Camera<'d, I2S: DmaEligible> {
```
The `I2S` generic parameter should be erased, like this https://github.com/esp-rs/esp-hal/blob/7c159b6070b9025903de7d24ac600839625a191f/esp-hal/src/i2s/parallel.rs#L237-L244

Also, do the guard thing.

rursprung commented on esp-hal/src/i2s/master/camera.rs:
```diff
@@ -0,0 +1,470 @@
+        i2s: impl Peripheral<P = I2S> + 'd,
+        config: Config,
+        channel: impl Peripheral<P = CH> + 'd,
+    ) -> Self {
```
note that this returns `Self` instead of `Result<Self, ConfigError>` because - at this point - there is no error which can happen. should i have an empty `ConfigError` enum just to fulfill the dev. guide? adding an enum entry later would anyway be a breaking change for consumers, so we can as well switch to a `Result` later if it turns out that there are errors

bugadani replied:
> should i have an empty ConfigError enum just to fulfill the dev. guide?

Yes please. Adding variants to `#[non_exhaustive]` enums is not considered a breaking change.

MabezDev commented:
What's the status here? Is this PR ready for testing?

rursprung commented:
> What's the status here? Is this PR ready for testing?

it needs rebasing (there were quite some changes upstream again, esp. related to peripherals) and addressing of review feedback. i don't have time right now, so anyone who wants to pick this up: please have a go! otherwise it might take several weeks (at least) until i get back to this.

Token-Thinker commented:
> > What's the status here? Is this PR ready for testing?
> 
> it needs rebasing (there were quite some changes upstream again, esp. related to peripherals) and addressing of review feedback. i don't have time right now, so anyone who wants to pick this up: please have a go! otherwise it might take several weeks (at least) until i get back to this.

I'm attempting rebasing and ummmmmmm yeahhhhh. i'll see how i can assit. finished the qa-test but still working on the `master/mod.rs` wish me luck on my learning journey haha

https://github.com/Token-Thinker/esp-hal/tree/rebasing-for-3219
